src/trust/chain.ts

```diff
@@ -82,13 +97,78 @@ export async function validateTrustChain(
     return shape.parse(decode(nextToken)).payload.jwks.keys;
   };
 
+  const x509TrustAnchorCertBase64 =
+    getTrustAnchorX509Certificate(trustAnchorEntity);
+
   // Iterate the chain and validate each element's signature against the public keys of its next
   // If there is no next, hence it's the end of the chain, and it must be verified by the Trust Anchor
-  return Promise.all(
-    chain
-      .map((token, i) => [token, selectKid(i), selectKeys(i)] as const)
-      .map((args) => verify(...args))
-  );
+  const validationPromises = chain.map(async (tokenString, i) => {
+    const kidFromTokenHeader = selectKid(i);
+    const signerJwks = selectKeys(i);
+
+    // Step 1: Verify JWT signature
+    const parsedToken = await verify(
+      tokenString,
+      kidFromTokenHeader,
+      signerJwks
+    );
+
+    // Step 2: X.509 Certificate Chain Validation
+    const jwkUsedForVerification = signerJwks.find(
+      (k) => k.kid === kidFromTokenHeader
+    );
+
+    if (!jwkUsedForVerification) {
+      throw new FederationError(
+        `JWK with kid '${kidFromTokenHeader}' was not found in signer's JWKS for token at index ${i}, though JWT verification passed.`,
+        { tokenIndex: i, kid: kidFromTokenHeader }
+      );
+    }
+
+    if (
+      !jwkUsedForVerification.x5c ||
+      jwkUsedForVerification.x5c.length === 0
+    ) {
+      throw new MissingX509CertsError(
+        `JWK with kid '${kidFromTokenHeader}' does not contain an X.509 certificate chain (x5c) for token at index ${i}.`
+      );
+    }
+
+    const originalX5cChain: string[] = jwkUsedForVerification.x5c;
+    let certChainToPassToNative = [...originalX5cChain];
+
+    // If the chain has more than one certificate AND
+    // the last certificate in the x5c chain is the same as the trust anchor,
+    // remove the anchor from the chain being passed, as it's supplied separately.
+    if (
+      certChainToPassToNative.length > 1 &&
+      certChainToPassToNative.at(-1) === x509TrustAnchorCertBase64
+    ) {
+      certChainToPassToNative.pop(); // Remove the last element
+    }
```

Comment:
What do you think about doing something like this?
```suggestion
const certChainBase64 =
   jwkUsedForVerification.x5c.length > 1 &&
   jwkUsedForVerification.x5c.at(-1) === x509TrustAnchorCertBase64
     ? jwkUsedForVerification.x5c.slice(0, -1)
     : jwkUsedForVerification.x5c;
```